control/discovery.py

```diff
@@ -0,0 +1,957 @@
+            nvme_tcp_data_pdu.data_length = b'\x10\x00\x00\x00'
+
+            # NVM Express
+            nvme_get_log_page_reply = NVMeGetLogPage()
+            nvme_get_log_page_reply.genctr = struct.pack('<Q', self_gen_cnt)
+            nvme_get_log_page_reply.numrec = struct.pack('<Q', len(listeners))
+
+            reply = pdu_reply.compose_reply() + nvme_tcp_data_pdu.compose_reply() + \
+                    nvme_get_log_page_reply.compose_short_reply()
+            try:
+                sock.sendall(reply)
+            except BrokenPipeError:
+                self.logger.debug("client disconnected unexpectedly.")
+                return -1, 0, 0
+        elif nvme_data_len > 16 and nvme_data_len % 1024 == 0:
+            # class Pdu and NVMeTcpDataPdu
+            pdu_and_nvme_pdu_len = 8 + 16
+
+            pdu_reply = Pdu()
+            pdu_reply.type = bytearray([NVME_TCP_PDU.C2H_DATA])
+            pdu_reply.specical_flag = b'\x0c'
+            pdu_reply.header_length = b'\x18'
+            pdu_reply.data_offset = b'\x18'
+            pdu_reply.packet_length = struct.pack('<I', pdu_and_nvme_pdu_len + nvme_data_len)
+
+            # NVMe/TCP Data PDU
+            nvme_tcp_data_pdu = NVMeTcpDataPdu()
+            nvme_tcp_data_pdu.cmd_id = struct.pack('<H', cmd_id)
+            nvme_tcp_data_pdu.data_length = struct.pack('<I', nvme_data_len)
+
+            # NVM Express
+            reply = pdu_reply.compose_reply() + nvme_tcp_data_pdu.compose_reply() + \
+                    log_page[0:nvme_data_len]
+            log_page = log_page[nvme_data_len:]
+            log_page_len -= nvme_data_len
+            try:
+                sock.sendall(reply)
+            except BrokenPipeError:
+                self.logger.debug("client disconnected unexpectedly.")
+                return -1, 0, 0
+        else:
+            self.logger.error(f"lenghth error. It need be 16 or n*1024")
+            return -1, 0, 0
+        self.logger.debug("reply get log page request.")
+        return 0, log_page, log_page_len
+
+
+    def nvmeof_tcp_connection(self, sock, addr):
+        self.logger.info(f"Accept new connection from {addr}")
+        # Some common variables
+        nvmeof_connect_data_hostid = ()
+        nvmeof_connect_data_cntlid = 0
+        nvmeof_connect_data_subnqn = ()
+        nvmeof_connect_data_hostnqn = ()
+        err = 0
+        sq_head_ptr = 0
+        log_page = bytearray()
+        log_page_len = 0
+        shutdown_notification = 0
+        shutdown_now = 0
+        global GLOBAL_CNLID
+        global GLOBAL_GEN_CNT
+        lock.acquire()
+        self_cnlid = GLOBAL_CNLID
+        self_gen_cnt = GLOBAL_GEN_CNT
+        GLOBAL_CNLID += 1
+        GLOBAL_GEN_CNT += 1
+        lock.release()
+
+        while True:
+            sq_head_ptr += 1
+            if sq_head_ptr > SQ_HEAD_MAX:
+                sq_head_ptr = 1
+            head = sock.recv(NVME_TCP_PDU_UNIT)
+            if not head:
+                time.sleep(0.01)
```

Comment:
Probably sleep is not required before the break if we are heading to socket close.